cpp/src/groupby/hash/groupby.cu

```diff
@@ -93,117 +94,128 @@ bool constexpr is_hash_aggregation(aggregation::Kind t)
     }
   }
-
-  auto get_flattened()
+  void Dispatch(cudf::detail::std_var_aggregation& agg) override
   {
-    return std::make_tuple(table_view(columns), std::move(agg_kinds), std::move(col_ids));
+    // TODO
   }
 };
 
-template <>
-void flatten_to_single_pass_aggregation::operator()<aggregation::COUNT_VALID>(
-  size_t i, column_view const& request_values, aggregation const& agg)
-{
-  insert_simple_aggs(i, request_values, agg);
-}
-template <>
-void flatten_to_single_pass_aggregation::operator()<aggregation::COUNT_ALL>(
-  size_t i, column_view const& request_values, aggregation const& agg)
-{
-  insert_simple_aggs(i, request_values, agg);
-}
-
-template <>
-void flatten_to_single_pass_aggregation::operator()<aggregation::MIN>(
-  size_t i, column_view const& request_values, aggregation const& agg)
-{
-  if (request_values.type().id() == type_id::STRING) {
-    insert_simple_aggs(i, request_values, {aggregation::ARGMIN});
-  } else {
-    insert_simple_aggs(i, request_values, agg);
-  }
-}
-
-template <>
-void flatten_to_single_pass_aggregation::operator()<aggregation::MAX>(
-  size_t i, column_view const& request_values, aggregation const& agg)
-{
-  if (request_values.type().id() == type_id::STRING) {
-    insert_simple_aggs(i, request_values, {aggregation::ARGMAX});
-  } else {
-    insert_simple_aggs(i, request_values, agg);
-  }
-}
-
-template <>
-void flatten_to_single_pass_aggregation::operator()<aggregation::MEAN>(
-  size_t i, column_view const& request_values, aggregation const& agg)
-{
-  auto mean_agg = static_cast<cudf::detail::mean_aggregation const&>(agg);
-  insert_simple_aggs(
-    i, request_values, static_cast<cudf::detail::compound_aggregation const&>(mean_agg));
-}
-
-template <>
-void flatten_to_single_pass_aggregation::operator()<aggregation::VARIANCE>(
-  size_t i, column_view const& request_values, aggregation const& agg)
-{
-  auto var_agg = static_cast<cudf::detail::std_var_aggregation const&>(agg);
-  insert_simple_aggs(
-    i, request_values, static_cast<cudf::detail::compound_aggregation const&>(var_agg));
-}
-
 // flatten aggs to filter in single pass aggs
 std::tuple<table_view, std::vector<aggregation::Kind>, std::vector<size_t>>
 flatten_single_pass_aggs(std::vector<aggregation_request> const& requests)
 {
-  flatten_to_single_pass_aggregation flattened{};
+  std::vector<column_view> columns;
+  std::vector<aggregation::Kind> agg_kinds;
+  std::vector<size_t> col_ids;
+  auto insert_agg = [&](size_t i, column_view const& request_values, aggregation::Kind k) {
+    agg_kinds.push_back(k);
+    columns.push_back(request_values);
+    col_ids.push_back(i);
+  };
+
   for (size_t i = 0; i < requests.size(); i++) {
     auto const& request = requests[i];
     auto const& agg_v   = request.aggregations;
 
     for (auto&& agg : agg_v) {
-      aggregation_dispatcher(agg->kind, flattened, i, request.values, *agg);
+      auto comp_agg = dynamic_cast<cudf::detail::compound_aggregation*>(agg.get());
```

Comment:
What if instead of querying if the `agg` is a `compound_aggregation` we just add `get_simple_aggregation` and `finalize` be members of the `aggregation` base class. For "simple" aggs, `finalize` is just a pass through function.

Reply:
Done.